### PR TITLE
Unwind info cleanups

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

File: crates/cranelift/src/compiler.rs [hunk 1]
```diff
@@ -261,6 +261,7 @@ impl wasmtime_environ::Compiler for Compiler {
             trampolines.push(builder.trampoline(*i, &func));
         }
         builder.align_text_to(CODE_SECTION_ALIGNMENT);
+        builder.unwind_info(&*self.isa);
```
Comment (alexcrichton) [inline crates/cranelift/src/compiler.rs:264]: I think this branch needs a rebase on `main`, but as a heads up when you do that `self.isa` is now also stored within the `ObjectBuilder` itself so it'll no longer need to be an argument here.

File: crates/cranelift/src/obj.rs [hunk 2]
```diff
@@ -101,6 +101,12 @@ pub struct ObjectBuilder<'a> {
     pending_relocations: Vec<(u64, &'a [Relocation])>,
     windows_unwind_info: Vec<RUNTIME_FUNCTION>,
     systemv_unwind_info: Vec<(u64, &'a systemv::UnwindInfo)>,
+
+    // The unwind info _must_ come directly after the text section. Our FDE's
+    // instructions are encoded to rely on this placement. We use this `bool`
+    // for debug assertions to ensure that we get the ordering correct.
+    #[cfg(debug_assertions)]
```
Comment (alexcrichton) [inline crates/cranelift/src/obj.rs:108]: FWIW I don't think it's worth doing the `debug_assertions` dance here, I'd just make everything `assert!` to unconditionally check.

File: crates/jit/src/unwind/systemv.rs [hunk 3]
```diff
@@ -25,6 +25,12 @@ impl UnwindRegistration {
         unwind_info: *mut u8,
         unwind_len: usize,
     ) -> Result<UnwindRegistration> {
+        debug_assert_eq!(
+            unwind_info as usize % 0x1000,
```
Comment (alexcrichton) [inline crates/jit/src/unwind/systemv.rs:29]: I think that this `0x1000` can probably be `region::page::size()` here in this crate

**Review by alexcrichton — COMMENTED** [review]
Looks good to me! Although this may change a fair bit when rebased on `main`

One thing I might recommend as well is to have another assertion when the object file is actually created that the `.eh_frame` and `.text` sections are indeed adjacent as assumed during construction?

**Comment by fitzgen** [comment]
Rebased and also addressed those comments.

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**